src/schemes/fluid/viscosity.jl

```diff
@@ -85,31 +85,77 @@ struct ArtificialViscosityMonaghan{ELTYPE}
+@doc raw"""
+    ViscosityMorris(; nu, epsilon=0.01)
+
+Viscosity by Moris (Moris et al. 1997).
+```math
+f_{ab} = \sum_w \frac{m_b(eta_a+eta_b)}{||r_{ab}||^2+(\epsilon h_{ab})^2}  \nabla W_{ab} \cdot r_{ab}\cdot v_{ab},
+```
+where ``\eta_a = \rho_a \nu`` and ``\eta_b = \rho_b \nu`` denote the dynamic viscosity
+of particle ``a`` and ``b`` respectively, and ``\nu`` is the kinematic viscosity.
+
+# Keywords
+- `nu`: Kinematic viscosity
+- `epsilon=0.01`: Parameter to prevent singularities
+
+## References
+- Joseph P. Morris, Patrick J. Fox, Yi Zhu.
+  "Modeling Low Reynolds Number Incompressible Flows Using SPH".
+  In: Journal of Computational Physics, Volume 136, Issue 1 (1997), pages 214--226.
+  [doi: doi.org/10.1006/jcph.1997.5776](https://doi.org/10.1006/jcph.1997.5776)
+- Georgios Fourtakas, Jose M. Dominguez, Renato Vacondio, Benedict D. Rogers.
+  "Local uniform stencil (LUST) boundary condition for arbitrary
+  3-D boundaries in parallel smoothed particle hydrodynamics (SPH) models".
+  In: Computers & Fluids, Volume 190 (2019), pages 346--361.
+"""
+struct ViscosityMorris{ELTYPE}
+    nu::ELTYPE
+    epsilon::ELTYPE
+
+    function ViscosityMorris(; nu, epsilon=0.01)
+        new{typeof(nu)}(nu, epsilon)
+    end
+end
+
+function kinematic_viscosity(system, viscosity::ViscosityMorris)
+    return viscosity.nu
+end
+
+@inline function (viscosity::Union{ArtificialViscosityMonaghan, ViscosityMorris})(particle_system,
+                                                                                  neighbor_system,
+                                                                                  v_particle_system,
+                                                                                  v_neighbor_system,
+                                                                                  particle,
+                                                                                  neighbor,
+                                                                                  pos_diff,
+                                                                                  distance,
+                                                                                  sound_speed,
+                                                                                  m_a, m_b,
+                                                                                  rho_mean)
     (; smoothing_length) = particle_system
 
+    rho_a = particle_density(v_particle_system, particle_system, particle)
+    rho_b = particle_density(v_neighbor_system, neighbor_system, neighbor)
+
     v_a = viscous_velocity(v_particle_system, particle_system, particle)
     v_b = viscous_velocity(v_neighbor_system, neighbor_system, neighbor)
     v_diff = v_a - v_b
 
     # v_ab ⋅ r_ab
     vr = dot(v_diff, pos_diff)
 
-    pi_ab = viscosity(sound_speed, vr, distance, rho_mean, smoothing_length)
+    grad_kernel = smoothing_kernel_grad(particle_system, pos_diff, distance)
```

Comment:
This is also available when calling this function. We should even get a considerable speedup when we stop unnecessarily computing the kernel gradient twice.